### ISSUE TITLE
Add suspicious DDoS IPs in deny Apache + log file + on option send email to server admin

### ISSUE THREAD

**Issue by pH-7** [pH7Software/pH7-Social-Dating-CMS#64]
Although DDoS attacks and request checks must normally be treated before PHP execution, here, in order to facilitate the task for newbies and also to continue the [Stop class](https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/master/_protected/framework/Security/DDoS/Stop.class.php#L21) and the requests from pH7CMS's users.

Add suspicious attacker IPs in the .htaccess file like below:

``` PHP
    $sHtaccess = file_get_contents('.htaccess');
    OPEN IP BAN LOG TOO
    ADD $sHtaccess .= "Deny from ".$sIp." \r\n"; into the .htaccess
       file_put_contents('.htaccess', $sHtaccess, FILE_APPEND);
    $sEmailText = t('DoS attack from %0% was blocked on %1%', $sIp, $sDate);

    if (config file email is enabled ) {
       Send Email
    }
```

**Comment by polynamaude** [comment]
I have a php script that could possibly be integrated into this function. It does a request to two "bad ips" list server on the web and output them over .htaccess as mod_rewrite rule to give a "F,L" (Forbidden and Last).

If you're interested just give me a note on the integration and I'll do as much as I can. I currently have a contract for a security threat management company so I could take some time on the job do to it.

**Comment by pH-7** [comment]
Your script looks a great idea indeed! That issue was an old one (and was planed for 2.0 release 😂).
Thank you for reminding me anyway 👍 Not sure how difficult it could be integrated... but that would be awesome!

**Comment by polynamaude** [comment]
Tell me a bit more about what you had planned at first and I'll do my part of the job of making that work in the framework you had planned to operate in.

As I understand it would be to write directly into the .htaccess file ?

**Comment by pH-7** [comment]
Hum, yes, the initial plan was just adding the suspicious attacker IPs into the .htaccess file from https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/6a5457a5f532b448f8526662487cd250bd036356/_protected/framework/Mvc/Controller/Controller.class.php#L331
If `if ($oDDoS->cookie() || $oDDoS->session()) {` is true, then it might be a suspicious request.

Now, I think we should be careful with the .htaccess file permissions. However, in most cases, it should be OK.

**Comment by polynamaude** [comment]
Well we've got two choice,
first one is to use (and modify) a .htaccess file to deny access to the website. We could use some text marker to indicate part of the file that is "managed" by the application.

option two, we create a easy processing option that returns a header 403/404 in case of suspected DDoS. The last option may give some independence from Apache but requires more computing power. There will always be some people using other server but Apache is still (probably) the most used Web server in the earth based universe.

I'll be pushing some code this monday - tuesday. Afterward I'll get on to this project.

**Comment by pH-7** [comment]
I think, I will vote for the second once...! I really like the idea! (although if it requires more computing  power).
With the version 1, .htaccess could become very big for some websites... and the list won't be updated, even if the backlisted IP isn't relevant anymore.

Let me know if you need any help ;),
Thanks!

**Comment by pH-7** [comment]
FYI, something that can already help alot has already been released thanks @polynamaude's work: https://github.com/pH7Software/pH7-Social-Dating-CMS/pull/209

All the new changes have been done and recently updated for [v15.1.7](https://github.com/pH7Software/pH7-Social-Dating-CMS/releases/tag/15.1.7) 🚀

**Comment by pH-7** [comment]
This has been implemented in different ways already, in [v15.1.7](https://ph7cms.com/ph7builder-15-1-7-released/) and has now been improved in [v17.1.8](https://github.com/pH7Software/pH7-Social-Dating-CMS/releases/tag/17.1.8).
